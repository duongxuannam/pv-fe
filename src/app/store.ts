import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit"
import { setupListeners } from "@reduxjs/toolkit/query"
import userReducer from "../features/user/userSlice"
import { devicesApi } from "../features/devices/devicesAPI"

export const store = configureStore({
  reducer: {
    user: userReducer,
    [devicesApi.reducerPath]: devicesApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(devicesApi.middleware),
})

setupListeners(store.dispatch)

export type AppStore = typeof store
export type AppDispatch = typeof store.dispatch
export type RootState = ReturnType<typeof store.getState>
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>
