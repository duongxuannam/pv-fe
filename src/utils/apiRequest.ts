import { BaseQueryFn } from "@reduxjs/toolkit/dist/query"
import axios, { AxiosError, AxiosRequestConfig, AxiosResponse } from "axios"

export const baseURL = "https://pv-be.onrender.com/"

const axiosInstance = axios.create({
  // baseURL: baseURL,
  timeout: 8000,
  headers: { Accept: 'application/json', },
})

export const setAuthenticationHeader = (jwt: string) => {
  axiosInstance.defaults.headers.common['Authorization'] = jwt
}

export const clearAuthenticationHeader = () => {
  axiosInstance.defaults.headers.common['Authorization'] = ''
}




axiosInstance.interceptors.response.use(
  function (response: AxiosResponse) {
    // co the format data truoc khi tra ve o day;
    return response
  },
  function (error: AxiosError) {
    console.log('no vo cay har', error)
    /// handle some error here
    if (error.code === "401") {
      // token dang nhap het han logout o day
      clearAuthenticationHeader()
    }
    return Promise.reject(error)
  },
)


export const axiosBaseQuery =
  (
    { baseUrl }: { baseUrl: string } = { baseUrl: '' }
  ): BaseQueryFn<
    {
      url: string
      method: AxiosRequestConfig['method']
      data?: AxiosRequestConfig['data']
      params?: AxiosRequestConfig['params']
    },
    unknown,
    unknown
  > =>
    async ({ url, method, data, params }) => {

      try {
        const result = await axiosInstance({ url: baseUrl + url, method, data, params })
        return { data: result.data }
      } catch (axiosError) {
        let err = axiosError as AxiosError
        return {
          error: {
            status: err.response?.status,
            data: err.response?.data || err.message,
          },
        }
      }
    }

export default axiosInstance
