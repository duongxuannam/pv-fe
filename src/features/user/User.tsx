import { useEffect } from "react"
import { useAppSelector, useAppDispatch } from "../../app/hooks"
import { fetchUserRedux, selectUser } from "./userSlice"

const url =
  "https://scontent.fsgn2-7.fna.fbcdn.net/v/t1.18169-1/11193438_402423169882782_2597021278587966343_n.jpg?stp=c0.2.40.40a_cp0_dst-jpg_p40x40&_nc_cat=108&ccb=1-7&_nc_sid=7206a8&_nc_ohc=iCEUNNQblrkAX91dsZP&_nc_ht=scontent.fsgn2-7.fna&oh=00_AfClD5sQ-_x_MYzw8cqGVumh1oerFns1BnnwjxMTq8Bhtg&oe=64EC1BA8"

export function User() {
  const dispatch = useAppDispatch()
  const user = useAppSelector(selectUser)

  useEffect(() => {
    const getData = () => {
      dispatch(fetchUserRedux())
    }
    getData()
  }, [dispatch])
  return (
    <header className="h-16 bg-sky-500  flex justify-end items-center">
      <h1
        data-testid="name"
        className="text-lg font-bold underline text-zinc-950"
      >
        {user.name}
      </h1>
      <img
        className="w-10 h-10 rounded-full mx-5"
        alt="Rounded avatar"
        src={url}
      />
    </header>
  )
}
