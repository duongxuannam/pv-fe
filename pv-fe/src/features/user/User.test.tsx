import { render, screen, waitFor } from "@testing-library/react"
import { Provider } from "react-redux"
import { rest } from "msw"
import { setupServer } from "msw/node"
import { store } from "../../app/store"
import { User } from "./User"
import { baseURL } from "../../utils/apiRequest"

const mockData = {
  id: "1",
  name: "Xuan Nam",
  title: "dev",
}

export const handlers = [
  rest.get(`${baseURL}users`, (req, res, ctx) => {
    return res(ctx.json(mockData))
  }),
]

const server = setupServer(...handlers)

// Enable API mocking before tests.
beforeAll(() => server.listen())

// Reset any runtime request handlers we may add during the tests.
afterEach(() => server.resetHandlers())

// Disable API mocking after the tests are done.
afterAll(() => server.close())

test("Check snapshot", async () => {
  const view = render(
    <Provider store={store}>
      <User />
    </Provider>,
  )
  expect(view).toMatchSnapshot()

  await waitFor(() => {
    expect(screen.getByTestId("name")).toHaveTextContent(mockData.name)
  })
})
