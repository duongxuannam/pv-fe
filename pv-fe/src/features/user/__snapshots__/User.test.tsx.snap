// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Check snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="h-16 bg-sky-500  flex justify-end items-center"
      >
        <h1
          class="text-lg font-bold underline text-zinc-950"
          data-testid="name"
        />
        <img
          alt="Rounded avatar"
          class="w-10 h-10 rounded-full mx-5"
          src="https://scontent.fsgn2-7.fna.fbcdn.net/v/t1.18169-1/11193438_402423169882782_2597021278587966343_n.jpg?stp=c0.2.40.40a_cp0_dst-jpg_p40x40&_nc_cat=108&ccb=1-7&_nc_sid=7206a8&_nc_ohc=iCEUNNQblrkAX91dsZP&_nc_ht=scontent.fsgn2-7.fna&oh=00_AfClD5sQ-_x_MYzw8cqGVumh1oerFns1BnnwjxMTq8Bhtg&oe=64EC1BA8"
        />
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="h-16 bg-sky-500  flex justify-end items-center"
    >
      <h1
        class="text-lg font-bold underline text-zinc-950"
        data-testid="name"
      />
      <img
        alt="Rounded avatar"
        class="w-10 h-10 rounded-full mx-5"
        src="https://scontent.fsgn2-7.fna.fbcdn.net/v/t1.18169-1/11193438_402423169882782_2597021278587966343_n.jpg?stp=c0.2.40.40a_cp0_dst-jpg_p40x40&_nc_cat=108&ccb=1-7&_nc_sid=7206a8&_nc_ohc=iCEUNNQblrkAX91dsZP&_nc_ht=scontent.fsgn2-7.fna&oh=00_AfClD5sQ-_x_MYzw8cqGVumh1oerFns1BnnwjxMTq8Bhtg&oe=64EC1BA8"
      />
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
